Add logic to SET enable_domain_typmod guc for domains having typmod enabled

rishabhtanwar29 opened the pull request:
During pg_dump we loose the setting of enable_domain_typmod
GUC and hence restored domains show incorrect behaviour.

1. Fix this by dumping enable_domain_typmod guc setting
   for the domains which supports domain typmod.
2. Also, replace unnecessary sql_dialect check in function 
   `coerce_to_domain` with typname check for some predefined
    domains. T-SQL allows typmod for the following domains:
    sys.nvarchar, sys.nchar, sys.varbinary, sys.binary, sys.decimal

Task: BABEL-3107
Signed-off-by: Rishabh Tanwar <ritanwar@amazon.com>

### Check List
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

rishabhtanwar29 committed:
Allow typmod to some specific TSQL domain types

During pg_dump we loose the setting of enable_domain_typmod
GUC and hence restored domains show incorrect behaviour.

Fix this by allowing typmod for some predefined domains
instead of just relying on guc setting. TSQL allows typmod
for the following domain types:

sys.nvarchar, sys.nchar, sys.varbinary, sys.binary,
sys.decimal

Signed-off-by: rishabhtanwar29 <rishabhtanwar29@gmail.com>

2jungkook commented on src/backend/commands/typecmds.c:
```diff
@@ -868,7 +868,13 @@ DefineDomain(CreateDomainStmt *stmt)
 
 	/* Only allow typmods if GUC is set, used by babelfishpg_tsql extension */
-	if (enable_domain_typmod)
+	if (enable_domain_typmod ||
```
It will work in a short-term, i.e., MVU for sys catalog, however a more desirable solution would be to modify pg_dump when considering user-defined domains, e.g.,
```
SET enable_domain_typmod = TRUE;
CREATE DOMAIN user_defined_domain_with_typmod AS user_defined_type;
RESET enable_domain_typmod;

SET enable_domain_typmod = FALSE;
CREATE DOMAIN user_defined_domain_without_typmod AS user_defined_type;
RESET enable_domain_typmod;
```

If we are able to detect whether a domain has typmod enabled during pg_dump, then we can generate the same SQL as above.

rishabhtanwar29 replied:
Added above logic in pg_dump.

2jungkook reviewed (commented):
Not specific to this PR, but I think we need a way to trigger Github actions for the extension package even the change is only for the engine.

rishabhtanwar29 committed:
Add logic to SET enable_domain_typmod guc for domains having typmod enabled

During pg_dump we loose the setting of enable_domain_typmod
GUC and hence restored domains show incorrect behaviour.

Fix this by also dumping enable_domain_typmod guc setting
for the domains which supports domain typmod.

Signed-off-by: Rishabh Tanwar <ritanwar@amazon.com>

2jungkook commented on src/bin/pg_dump/pg_dump.c:
```diff
@@ -11371,6 +11382,10 @@ dumpDomain(Archive *fout, const TypeInfo *tyinfo)
 
+	if (typsupporttypmod)
+		appendPQExpBuffer(q,
+						"SET enable_domain_typmod = FALSE;\n");
```
Let's restore the previous value instead of putting a hard-coded value.

rishabhtanwar29 replied:
Fixed!

rishabhtanwar29 committed:
RESET the value of enable_domain_typmod

Signed-off-by: Rishabh Tanwar <ritanwar@amazon.com>

2jungkook reviewed (approved):
Let's use `mvu-dev` branch.
